runtime/src/main/java/org/corfudb/runtime/collections/TxnContext.java

```diff
@@ -0,0 +1,517 @@
+            decision |= 0x01;
+        }
+        switch (decision) {
+            case 0: // commit() on a normal write-only transaction!
+                log.trace("commit() applying {} operations", operations.size());
+                txBeginInternal();
+                operations.forEach(Runnable::run);
+                operations.clear();
+                txnType = TxnType.writeOnly;
+                break;
+            case 1: // commit() called without a corfu transaction or buffered writes!
+                log.warn("commit called on an empty transaction!");
+                break;
+            case 2: // commit() called but corfu txn already begun for reads
+                log.trace("commit() applying pending {} operations after reads", operations.size());
+                operations.forEach(Runnable::run);
+                operations.clear();
+                txnType = TxnType.writeAndReads;
+                break;
+            case 3: // commit called in a read-only transaction.
+                if (txnType == TxnType.readOnly) {
+                    log.trace("Ending a read-only transaction");
+                } else if (txnType == TxnType.writeAndReads) {
+                    log.trace("Ending a dirty-read transaction!");
+                } else { // Attempting to commit a corfu txn started outside this scope!
+                    log.error("commit() on a transaction started outside scope of this thread!");
+                    throw new IllegalStateException(
+                            "commit() invoked but transaction not started in this thread!");
+                }
+                break;
+            case 4:
+                log.trace("Dirty reads invoked. Applying {} in thread", operations.size());
+                tablesInTxn.add(tableBeingRead);
+                txBeginInternal();
+                operations.forEach(Runnable::run);
+                operations.clear();
+                txnType = TxnType.writeAndReads;
+                break;
+            case 5: // Normal path for a starting a transaction to do queries.
+                log.trace("Starting a transaction for reading data");
+                tablesInTxn.add(tableBeingRead);
+                txBeginInternal();
+                txnType = TxnType.readOnly;
+                break;
+            case 6:
+                log.trace("More dirty reads. Applying {}", operations.size());
+                tablesInTxn.add(tableBeingRead);
+                operations.forEach(Runnable::run);
+                txnType = TxnType.writeAndReads;
+                operations.clear();
+                break;
+            case 7:
+                tablesInTxn.add(tableBeingRead);
+                if (txnType != TxnType.writeAndReads) {
+                    log.trace("More reads in same transaction");
+                    txnType = TxnType.readOnly;
+                } else {
+                    log.trace("More reads in transaction that had writes before");
```

Comment:
what are we trying to differentiate in these if-else?  Somehow it is not clear to me

Reply:
added a comment. we do not want to mark a transaction as read-only if there have been dirty writes before.
consider this case: 
write
read < dirty read - 
read < case 7 else part here